modules/@apostrophecms/ui/ui/apos/components/AposFile.vue

```diff
@@ -0,0 +1,207 @@
+      @dragleave="dragging = false"
+    >
+      <p class="apos-file-instructions">
+        <template v-if="dragging">
+          <cloud-upload-icon :size="38" />
+        </template>
+        <AposSpinner v-else-if="uploading" />
+        <template v-else>
+          <paperclip-icon :size="14" class="apos-file-icon" />
+          {{ messages.primary }}&nbsp;
+          <span class="apos-file-highlight" v-if="messages.highlighted">
+            {{ messages.highlighted }}
+          </span>
+        </template>
+      </p>
+      <input
+        type="file"
+        class="apos-sr-only"
+        :disabled="disabled || limitReached"
+        @input="uploadFile"
+        :accept="allowedExtensions"
+      >
+    </label>
+    <div v-if="filesOrAttachment.length" class="apos-file-files">
+      <AposSlatList
+        :value="filesOrAttachment"
+        @input="update"
+        :disabled="readOnly"
+      />
+    </div>
+  </div>
+</template>
+
+<script>
+export default {
+  props: {
+    uploading: {
+      type: Boolean,
+      default: false
+    },
+    disabled: {
+      type: Boolean,
+      default: false
+    },
+    limit: {
+      type: Number,
+      default: 1
+    },
+    attachment: {
+      type: Object,
+      default: null
+    },
+    allowedExtensions: {
+      type: String,
+      default: '*'
+    },
+    readOnly: {
+      type: Boolean,
+      default: false
+    },
+    def: {
+      type: String,
+      default: null
+    }
+  },
+  emits: [ 'upload-file', 'update' ],
+  data () {
+    return {
+      selectedFiles: [],
+      dragging: false
+    };
+  },
+  computed: {
+    limitReached () {
+      return this.filesOrAttachment.length >= this.limit;
+    },
+    filesOrAttachment () {
+      if (!this.selectedFiles.length && !this.attachment) {
+        return [];
+      }
+
+      return this.selectedFiles.length
+        ? this.selectedFiles
+        : [ this.attachment ];
+    },
+    messages () {
+      const msgs = {
+        primary: 'Drop a file here or',
+        highlighted: 'click to open the file explorer'
+      };
+      if (this.disabled) {
+        msgs.primary = 'Field is disabled';
+        msgs.highlighted = '';
+      }
+      if (this.limitReached) {
+        msgs.primary = 'Attachment limit reached';
+        msgs.highlighted = '';
+      }
+      return msgs;
+    }
+  },
+  methods: {
+    uploadFile ({ target, dataTransfer }) {
+      this.dragging = false;
+      const extensionRegex = /(?:\.([^.]+))?$/;
+
+      const files = target.files ? target.files : (dataTransfer.files || []);
+      const filteredFiles = Array.from(files).filter((file, i) => {
+        const [ extension ] = extensionRegex.exec(file.name);
+        return (i + 1 <= this.limit) && this.allowedExtensions.includes(extension);
+      });
+
+      this.selectedFiles = filteredFiles.map((file, i) => {
+        const [ _, extension ] = extensionRegex.exec((file.name));
+        const fakeId = `${i + 1}-${file.name}`;
```

Comment:
Because AposSlatList and AposSlat need an `_id`, that was the most confusing part for me because these components seem to be able to manage multiple items that must be attachments but in our case we manage only one, so the update method with items for example is confusing. This fakeId allowed me to display the selected file and allow to remove it without modifying any code from `AposSlatList` or `AposSlat`, I didn't want to take any side effect risk. (and they seem complex enough to avoid that).

Reply:
😮 I see. And the attachment input had a real _id because it uploaded the file to the attachments collection.